### ISSUE TITLE
[Impeller] blur coverage computations are still incorrect.

### ISSUE THREAD

**Issue by jonahwilliams** [flutter/flutter#145059]
Perhaps we're not expanding the bounds enough? Look at the bottom of the screen

https://github.com/flutter/flutter/assets/8975114/c5f0a7bf-62bd-4e43-842b-849f96ccaa55

Example app:

```dart
import 'dart:ui';

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: MyHomePage(title: 'Scrolling ios blur 3.19.1'),
    );
  }
}

class MyHomePage extends StatelessWidget {
  const MyHomePage({
    super.key,
    required this.title,
  });

  final String title;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      extendBody: true,
      extendBodyBehindAppBar: true,
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: Text(title),
      ),
      body: ListView.separated(
        itemCount: 20,
        itemBuilder: (_, i) => ImageFiltered(
            imageFilter: ImageFilter.blur(sigmaX: 25, sigmaY: 25),
            child: Image.network(
              'https://cdn.pixabay.com/photo/2018/03/31/19/29/schnitzel-3279045_1280.jpg',
              height: 200,
              fit: BoxFit.cover,
            ),
        ),
        separatorBuilder: (_, __) => const SizedBox(height: 80),
      ),
    );
  }
}
```

**Comment by jamesblasco** [comment]
Could it be related to this? https://github.com/flutter/flutter/issues/144985 We had these problems ourselves when using ImageFilter.blur in different scenarios

**Comment by jonahwilliams** [comment]
@jamesblasco I mean its potentially related since they are both blurs, could be that we are computing the coverage wrong which is impacting both places.

## Skia

If I run the Skia example, the blur goes offscreen much further instead of being cut-off. I need to get some frame captures to confirm what is going on.

![flutter_03](https://github.com/flutter/flutter/assets/8975114/ebac47ef-9488-4898-bc22-1ccd63188523)

**Comment by gaaclarke** [comment]
This seems to me a coverage bug.  The point at which it stops rendering is about the point where the coverage of sigma=0 is going off the screen.  It doesn't seem to be accounting for the blur radius.  We do have some unit tests for this sort of thing (GaussianBlurFilterContentsTest.CoverageWithSigma).  I'm not sure what path we would be going down to mess up with coverage.

**Comment by jonahwilliams** [comment]
Could also be an issue with the coverage hint value:

![image](https://github.com/flutter/flutter/assets/8975114/27097d1a-603b-4364-bae9-ae9512e5e833)


Looks like this expands offscreen a bit for width but not as much for height. though I'm sort of guessing since I don't know the input rect

**Comment by jonahwilliams** [comment]
I don't think this is a (recent) regression, but its good to know that we're not out of the woods on our coverage problems yet.

**Comment by gaaclarke** [comment]
If you put a print in `impeller::GaussianBlurFilterContents::GetFilterCoverage` it can be seen that the cell that isn't being drawn isn't even getting that far.  We just get the calls for the other 3 cells:
```
(impeller::TRect<float>)  (left_ = -79, top_ = -403.309387, right_ = 829, bottom_ = 154.690613)
(impeller::TRect<float>)  (left_ = -79, top_ = -403.309387, right_ = 829, bottom_ = 154.690613)
(impeller::TRect<float>)  (left_ = -79, top_ = -403.309387, right_ = 829, bottom_ = 154.690613)
(impeller::TRect<float>)  (left_ = -79, top_ = 156.690613, right_ = 829, bottom_ = 714.690613)
(impeller::TRect<float>)  (left_ = -79, top_ = 156.690613, right_ = 829, bottom_ = 714.690613)
(impeller::TRect<float>)  (left_ = -79, top_ = 156.690613, right_ = 829, bottom_ = 714.690613)
(impeller::TRect<float>)  (left_ = -79, top_ = 716.690613, right_ = 829, bottom_ = 1274.69067)
(impeller::TRect<float>)  (left_ = -79, top_ = 716.690613, right_ = 829, bottom_ = 1274.69067)
(impeller::TRect<float>)  (left_ = -79, top_ = 716.690613, right_ = 829, bottom_ = 1274.69067)
```
It's surprising that it omits that element without querying its coverage.

**Comment by jonahwilliams** [comment]
I think that is culling happening at the framework level. Originally I thought that was the bug, but instead I think its the white gap that appears - this looks substantially different with Skia - the blur continues to extend to the edge of the screen while the impeller version is cut off.

**Comment by gaaclarke** [comment]
Ah yea, culling in the framework would explain it.  I doubt the framework is calculating the blur radius exactly the same as the engine and it doesn't look like we are communicating that calculation in anyway to it.

The skia version is different, still wrong, just wrong in a different way.

https://github.com/flutter/flutter/assets/30870216/37f2c43d-d024-44dd-8840-884907000c68

**Comment by gaaclarke** [comment]
@jonahwilliams do you know how this is supposed to work?  The weird thing is that if I use a matrix filter that scales the image 1.5, the bug doesn't show up.  If it was a matter of the framework not having the correct information for culling, it seems like it works for the matrix filter but not the blur for some reason.

**Comment by gaaclarke** [comment]
> The framework culling is WAI. We only use layout sizes and if you paint outside of bounds then YMMV

So, is there somewhere in the framework that is looking at the image filter, seeing a matrix and expanding the layout size to accommodate that?  And if so, do we need to do the same for the blur?
